### ISSUE TITLE
plt.pcolormesh stopped working with Masked Arrays

### ISSUE THREAD

**Issue by divenex** [matplotlib/matplotlib#9720]
plt.pcolormesh does not work any more with Numpy Masked Arrays, as of Matplotlib 2.1.0

    import matplotlib.pyplot as plt
    import numpy as np

    x = np.linspace(0,1,10) 
    xx, yy = np.meshgrid(x, x)
    xx = np.ma.masked_array(xx, mask=xx>0.5)
    yy = np.ma.masked_array(yy, mask=xx>0.5)
    plt.pcolormesh(xx, yy, xx*yy)

produces the error

    File "Miniconda3\lib\site-packages\matplotlib\pyplot.py", line 3167, in pcolormesh ret = ax.pcolormesh(*args, **kwargs)
    File "Miniconda3\lib\site-packages\matplotlib\__init__.py", line 1710, in inner return func(ax, *args, **kwargs)
    File "Miniconda3\lib\site-packages\matplotlib\axes\_axes.py", line 5636, in pcolormesh
    coords = np.column_stack((X, Y)).astype(float, copy=False)
    TypeError: astype() got an unexpected keyword argument 'copy'

The example works as expected without using Numpy Masked Arrays. And indeed one can check that the `np.astype` does not accept the keyword `copy` for Masked Arrays.

This problem seems related to #9577

**Matplotlib version**

  * Operating system: Tested on both Windows and Ubuntu
  * Matplotlib version: 2.1.0
  * Matplotlib backend: Qt5Agg
  * Python version: 3.6.3
   * Matplotlib installed via Anaconda

**Comment by jklymak** [comment]
Thanks for the bug report...  Looks like #8801 caused this as well as what I fixed in #9168.

**Comment by jklymak** [comment]
Fixed in #9722

**Comment by efiring** [comment]
Unlike pcolor, pcolormesh isn't *designed* to work with masked arrays for x and y--they are supposed to be supported only for the c array:
```
       *C* may be a masked array, but *X* and *Y* may not.  Masked
        array support is implemented via *cmap* and *norm*; in
        contrast, :func:`~matplotlib.pyplot.pcolor` simply does not
        draw quadrilaterals with masked colors or vertices.
```
So I think the real bug is that we are not properly checking for the existence of masked or invalid points in the X and Y inputs; if found, they should trigger an exception.  Otherwise, I'm pretty sure that the masks will be discarded silently further down the line, which is not what we want.

**Comment by jklymak** [comment]
Oh, fair enough!  What I did indeed silently drops the mask from the x and y and just uses the underlying values.  

So, 

  1. catch masked arrays as TypeError 
  2. silently ignore (or log) the mask.

**Comment by jklymak** [comment]
Just to be clear, 1. is unfortunately an API change to bring the code in line w the docs.  2. keeps the API the same, but allows it to contravene the docs.

**Comment by efiring** [comment]
3. If X or Y are masked arrays, see if they have masked points; if they do, raise a ValueError, otherwise use asarray to convert them to unmasked arrays.  Possible addition: first use safe_masked_invalid, so that any nans or infs will also be caught early.

I would not support option 2, so I think it is between 1 and 3, or some minor variation on these.  Originally we considered it the responsibility of the user to pay attention to the docs and to ensure they supplied sensible and supported inputs; increasingly we have been putting in validation steps. Options 1 and 3 are simply two such approaches, with 3 being more complex but more user-friendly.

I'm OK with either; it might make sense to start with 1.  It's simple and consistent with the docstring. I suppose the downside is that it might break some existing code that was relying on the previous behavior, supplying masked arrays that happen to have valid values behind the mask.

Option 1 leaves the documented API intact, but breaks the effective API; Option 3 changes the documented API, but essentially leaves the effective API.

**Comment by jklymak** [comment]
I’m fine with 3 as well. I’ll hold off on a new PR in cases someone else wants to chime in.

**Comment by anntzer** [comment]
I think the doc is very clear that X and Y cannot be masked so I would just error out in that case... (so 1)
Alternatively, if we *do* want to support masked arrays then I think we should do so properly (which I think means masking out any quadrilateral where X or Y or C is masked).

**Comment by jklymak** [comment]
What about catching invalids?

**Comment by efiring** [comment]
@anntzer, remember, for the normal case (not Gouraud shading, which I think should be moved to a different function entirely) pcolormesh X and Y are boundaries, C is centers.  I don't see any rational way to support masked X and Y, or any compelling reason to try to do so.  Masked C is no problem, and is already handled properly.

**Comment by anntzer** [comment]
When I wrote "any quadrilateral where X or Y or C is masked", this should really read "any quadrilated that has either at least one vertex with at least one masked coordinate (X or Y), or a masked value (C)".

**Comment by divenex** [comment]
Agreed with being consistent with the documentation. My bad for relying on the old behaviour in my code. Indeed having only C supporting Masked Arrays is enough.

**Comment by divenex** [comment]
There is only one problem with the current implementation, namely that plot ranges are by default defined by the X and Y coordinates, which are not masked. This implies that in general one can have large empty areas of invisible masked rectangles in the plot area.

**Comment by efiring** [comment]
I think the solution in that case is for the user to set the xlim and ylim as desired.

**Comment by jklymak** [comment]
Yes.  If the user supplied x and y out to certain values despite a bunch of masked data,  presumably they have some need to indicate that those values are empty.

**Comment by divenex** [comment]
To give some context, [here](http://blog.sdss.org/wp-content/uploads/2015/04/fig2_bundy_etal_2015.png) is the type of application I have in mind. Indeed requesting the user to set the limits is acceptable, and this is the way I am doing it. But it would seem natural to automatically restrict the plotting limits to the edges of the non-masked rectangles, in the very same way as `plt.imshow` works with Masked Arrays. In fact I cannot think of an obvious user case where it is useful to display an extra empty border containing masked rectangles.

**Comment by jklymak** [comment]
I do it all the time when I grid data onto an even grid.  Sometimes the data isn't there, so it is masked, but it still gets displayed to compare w/ the data that as data there.  Otherwise all your data limits are different which can be misleading.  

Anyway, what you say could be done at the edges, but what about masked x, y in the interior?

**Comment by divenex** [comment]
Holes in the interior are OK and nothing needs to change there.

**Comment by divenex** [comment]
But note that the current behaviour of `plt.pcolormesh`, with Masked Arrays, is inconsistent with the one of `plt.imshow`. So one of the two must be non-optimal...

**Comment by jklymak** [comment]
`imshow` doesn't take values for `x` or `y`.  The issue we are addressing here is masked `x` and masked `y`.

**Comment by divenex** [comment]
Indeed, the fact that `x` and `y` do not accept masked arrays in `plt.pcolormesh`, now requires the user to set the limits by hand to avoid showing empty borders. I now understand that this is the intended behaviour, which is OK with me.

**Comment by divenex** [comment]
In fact I retract my statement that `plt.pcolormesh` is inconsistent with `plt.imshow` on this. They are actually consistent, so neither should change.

**Comment by tacaswell** [comment]
Closed by #9723
